numpyro/contrib/control_flow/scan.py

```diff
@@ -29,14 +29,20 @@ def tree_flatten(self):
                     # scanned sites have stop field because we trace them inside a block handler
                     elif key != 'stop':
                         aux_trace[name][key] = site[key]
-        return (trace,), aux_trace
+        # keep the site order information because in JAX, flatten and unflatten do not preserve
+        # the order of keys in a dict
+        site_names = list(trace.keys())
+        return (trace,), (aux_trace, site_names)
 
     @classmethod
     def tree_unflatten(cls, aux_data, children):
+        aux_trace, site_names = aux_data
         trace, = children
-        for name, site in trace.items():
-            site.update(aux_data[name])
-        return cls(trace)
+        trace_with_aux = {}
+        for name in site_names:
+            trace[name].update(aux_trace[name])
+            trace_with_aux[name] = trace[name]
+        return cls(trace_with_aux)
```

Comment:
I changed the logic here a bit to return a trace with sites having the same order as the original trace. No performance gain. This only makes it easier to debug (when I printed out the trace, the order of the sites was a bit surprising to me). :)